### PR TITLE
Started es6 migration and Rollup integration

### PR DESCRIPTION
**Pull request by hjemmel**
This commit only has:
- Started to migrate to es6
- Rollup for generation
- minified version added

### CODE REVIEW

**Comment by deitch** [comment]
Heh, I was _sure_ you would start with travis! No big deal, though, we can start here.

**Comment by deitch** [comment]
This is one serious job. I am impressed. 

1. Can you describe here how the build/run/test/import structure changes? And does the include process work any differently? 
2. As far as I can tell, `package.json` still points at `lib/searchjs.js`, so `npm install ... ` should "just work". Does it change in the browser?
3. We don't actually need babel or any transform to work in node, let alone in browsers that have es6 support, so do we want to compile/transpile it? I think it would be better as native es6 in node at least. For that matter, most modern browsers support mode of es6. I would be quite comfortable bumping major version to 1.0.0 and saying that 1.0.0 is supported only in es6-supportive browsers?

**Comment by hjemmel** [comment]
Hi @deitch,

1) `npm install` and `npm test` work as before. the only extra work is run `npm run build` to actually build the library, it generates a min version as well. The library usage remaining the same cuz I have used umd as build format and the name in rollup config is SEARCHJS (same you used to export to use in the browser).
I will introduce later some watchers to run the tests automatically once you change the library for example.

2) everything works as before just the build step I said before.

3) I made that way cuz of compatibility purposes and I think node does not es6 import style if I am not mistaken. If I try to run the tests without compiling/transpile first it will cause an error: `SyntaxError: Unexpected token import`. That's why I put the parameter `--compilers js:babel-core/register` in mocha.

Btw I saw an error here: tests should refer _require("../**src**/searchjs");_ and not _../**lib**/searchjs_ in that away later on when I configure the code coverage it will look at the correct code and not the compiled one.

but as you can see it works perfectly in browser and node!

**Comment by hjemmel** [comment]
just made another commit regarding the tests that I mentioned.

**Comment by deitch** [comment]
Thanks. I will try it out.

> compatibility purposes

With semver in place, I don't mind not being backwards compatible. But as you said:

> node does not es6 import style

Sort of does. Depends where and how. It has experimental support; see [docs](https://nodejs.org/api/esm.html) and [blog post](https://medium.com/the-node-js-collection/an-update-on-es6-modules-in-node-js-42c958b890c). Either way, yes, it is only experimental. But that is the only real sticking point.

As for the browsers, most support es6 at this point, but again, import is an issue, so you would have to rollup or similar.

**Comment by hjemmel** [comment]
Give it a try and let me know if we can improve something, I will introduce more npm scripts to run tests, serve a demo page and other things later on. Same with codecov and travis :)

**Comment by deitch** [comment]
Yeah, let's get this merged in, then PR for travis+codecov, then PRs for other stuff.

**Comment by deitch** [comment]
Mocha's `--compilers` flag is deprecated and subject to removal. See [here](https://github.com/mochajs/mocha/wiki/compilers-deprecation)

However, I could not figure out how to get it to work correctly with `--require`?

**Comment by deitch** [comment]
I got it. Comment coming.

File: package.json [hunk 1]
```diff
@@ -32,11 +32,20 @@
...
+    "rollup": "^0.60.7",
+    "rollup-plugin-babel": "^3.0.4",
+    "should": "^13.2.1",
+    "uglify-js": "^3.4.1"
   },
   "scripts": {
-    "test": "./node_modules/mocha/bin/mocha ./test/test.js --reporter spec"
+    "build": "NODE_ENV=prod rollup -c > ./lib/searchjs.js && npm run uglify",
+    "uglify": "uglifyjs ./lib/searchjs.js --compress --mangle --comments -o ./lib/searchjs.min.js",
+    "test": "NODE_ENV=test mocha --compilers js:babel-core/register ./test/test.js --reporter spec"
```
Comment (deitch) [inline package.json]: This should be changed to 

```json
  "test": "NODE_ENV=test mocha --require babel-register ./test/test.js --reporter spec"
```

File: package.json [hunk 2]
```diff
@@ -32,11 +32,20 @@
   },
   "bugs": "https://github.com/deitch/searchjs/issues",
   "devDependencies": {
-    "mocha": "^1.13.0",
-    "should": "^1.3.0"
+    "babel-core": "^6.26.3",
+    "babel-plugin-external-helpers": "^6.22.0",
+    "babel-preset-env": "^1.7.0",
+    "babel-preset-es2015": "^6.24.1",
```
Comment (deitch) [inline package.json]: Also needs:

```json
"babel-register": "^6.26.0",
```

More simply:

```
npm i --save-dev babel-register
```

**Comment by deitch** [comment]
This all is looking good. I have a few changes to suggest. I will put them in the "request changes."

Separately, I am questioning if we should be running tests on the _source_ `src/searchjs` or on the _compiled_ `lib/searchjs`. For now, we can leave it on source, but in most software development, we do run tests on compiled (i.e. deployed) artifacts, in case something strange gets triggered by the compilation process.

Since eventually the whole build stage will go away, I would leave it for now, but worth thinking of.

**Comment by hjemmel** [comment]
Just have added the required changes, I do agree that we should test on top of *lib/searchjs*. we can remove babel from mocha if we do that.

I will configure a source maps when I set up the codecov

**Comment by deitch** [comment]
Looking good! Everything is clean. One thing I missed is to bump the version. But I will merge it in and open a new PR for version number.

**Comment by deitch** [comment]
> Just have added the required changes, I do agree that we should test on top of lib/searchjs. we can remove babel from mocha if we do that.

Future one, I believe. And I think that we _should_ test against the source, else we can miss something there.

**Comment by deitch** [comment]
Thanks @hjemmel !

**Comment by deitch** [comment]
0.12.0 published.

**Pull request merged**